File: medium/300 (Longest increasing subsequence)/solution.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        List<Integer> ans = new ArrayList<>();
        for(int i = 0; i < nums.length; i++){
            if(i == 0) ans.add(nums[i]);
            else{
                if(nums[i] > ans.get(ans.size() - 1)) ans.add(nums[i]);
                else{
                    int indx = bs(ans,nums[i]);
                    ans.set(indx,nums[i]); 
                }
            }
        }
        return ans.size();
    }
    public int bs(List<Integer> ans, int c){
        int l = 0, r = ans.size() - 1;
        int mid = 0;
        int z = 0;
        while(l <= r){
            mid = l + (r - l) / 2;
            if(ans.get(mid) >= c){
                z = mid;
                r = mid - 1;
            }else{
                l = mid + 1;
            }
        }
        return z;
    }
}

File: medium/300 (Longest increasing subsequence)/solution.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        ans = [nums[0]]
        for i in range(1,len(nums)):
            if nums[i] > ans[len(ans) - 1]:
                ans.append(nums[i])
            else:
                indx = self.bs(ans,nums[i])
                ans[indx] = nums[i]
        return len(ans)
    def bs(self, ans: List[int],nums: int) -> int:
        l = 0
        r = len(ans) - 1
        z = 0
        while(l <= r):
            mid = int(l + (r - l) / 2)
            if ans[mid] >= nums:
                z = mid
                r = mid - 1
            else:
                l = mid + 1
        return z